### ISSUE TITLE
difference between CPU / CUDA while multiplying zero and nan

### ISSUE THREAD

**Issue by JosefJerabek** [arrayfire/arrayfire#2020]
`const af::array nan = af::constant(1.0, 1)) / (af::constant(0.0, 1);`
`const af::array zero = af::constant(1.0, 1) < 0.0;`
`const af::array product = zero * nan;`
`af_print(product);`

When linked with afcpu.lib gives this:
product
[1 1 1 1]
 -nan(ind)

When linked with afcuda.lib gives this:
product
[1 1 1 1]
    0.0000

**Comment by 9prady9** [comment]
This is happening only when one of the operands of * operator is a boolean type(may be non float types, I haven't checked other types). But If I do the following, it is working as expected.

```
const af::array nan = af::constant(1.0, 1)) / (af::constant(0.0, 1);
const af::array zero = af::constant(0.0, 1);
const af::array product = zero * nan;
af_print(product);
```

**Comment by 9prady9** [comment]
I have debugged it to some extent and given below are my findings.

* Adding `zero.eval()` after zero variable declaration gives correct results.
* Changing zero to be a `constant(0.0, 1, b8)` or constant array of any other type also works as expected.
* Seems to be related to `<` & `>` operators and scalars.
* Printed out the generated kernels from jit module and they checkout fine in terms of the operations they are doing.
* Verified if `__op` macros for any synctax errors in `src/backend/cuda/kernel/jit.cuh`. They look good.

The following specific set of statements is the only set I found to be breaking.
```
const af::array nan = af::constant(1.0, 1)) / (af::constant(0.0, 1);
const af::array zero = af::constant(1.0, 1) < 0.0;
const af::array product = zero * nan;
af_print(product);
```
@pavanky @umar456 Any hints on where I should be looking at for this problem ?
